2025 Year in Review (#2081)
withastro/astro.build: src/content/blog/year-in-review-2025.mdx

---
title: "2025 year in review"
description: "Revisit everything that happened in Astro last year: new feature releases, survey results, Astro Together, and more!"
# TODO: Update the date below when publishing
publishDate: "January 3, 2026"
authors:
  - shinya
  - sarah
# TODO: Update the images below when publishing
coverImage: "/src/content/blog/_images/year-in-review-2024/blog-post-2024-in-review.webp"

> socialImage: "/src/content/blog/_images/year-in-review-2024/og-2024-in-review.webp"

Author: These TODOs need to be completed before publishing this.

Reviewer: Done in 147a0afac891e2844cea272e5505ca7a1c42a580!

- Astro applies scoped styles using `data-astro-image` attributes to ensure images resize correctly without requiring manual CSS.

In addition, the following global styles are generated and applied by Astro to ensure that your images resize correctly:

```css
:where([data-astro-image]) {
  object-fit: var(--fit);
  object-position: var(--pos);
}
:where([data-astro-image='full-width']) {
  width: 100%;
}
:where([data-astro-image='constrained']) {
  max-width: 100%;
}
```

With the new responsive behavior, the necessary `srcset` and `sizes` values are generated automatically, and the appropriate styles for proper resizing are applied. Additionally, a `priority` property has been added to make it possible to load certain images with higher priority.

It is also worth noting that Astro [v5.7.0](https://astro.build/blog/astro-570/#svg-components) introduced support for importing SVG files as Astro components. This is another handy feature, so be sure to give it a try!

```astro
---
import Logo from './path/to/svg/file.svg';
---

<Logo width={64} height={64} fill="currentColor" />
```

> 

Reviewer 2: I know that the blog is already pretty long, but since we mention SVG imports here, a short section mentioning SVG optimisation added in [5.16.0](https://github.com/withastro/astro/pull/13880) could be very fitting here.

E.g. something like (with or without the code example:

````suggestion

Speaking of SVGs, Astro [`5.16.0`](https://astro.build/blog/astro-5160/#svg-optimization-with-svgo) added experimental support for automatic SVG optimization using SVGO.

```mjs
import { defineConfig } from 'astro/config';

export default defineConfig({
  experimental: {
    svgo: true,
  },
});
```

````

It is worth noting that inline scripts require custom hashes to be provided, and the `<ClientRouter />` component is not compatible with this feature. Be sure to test your CSP configuration using `astro build` and `astro preview`, as CSP is not enforced in development mode.

[Check out the official docs to learn more about experimental Content Security Policy in Astro](https://docs.astro.build/en/reference/experimental-flags/csp/).

If you are interested in learning through video, you may also want to check out the following:

<YouTube id="3r9ThIMYzsU" params="list=PL8Qn4kutqAEuDiRlQ7TD2lthxTKtrbQTc&index=1" posterQuality="high"/>

### Astro Docs MCP Server

> While 2025 has been widely described as the year of [vibe coding](https://x.com/karpathy/status/1886192184808149383), Astro has been steadily moving forward with support for AI tools as well. The Astro docs now include a new "[Build with AI](https://docs.astro.build/en/guides/build-with-ai/)" page, which provides information about `llms.txt` and offers tips for developing Astro projects with AI. The most extensively covered topic on that page is the Astro Docs MCP Server, which is introduced below.

Author: @trueberryless Thank you for the feedback! I was not aware that `llms.txt` might be removed. Based on reading that discussion, delucis’s reasoning seems reasonable to me, and it does look like there is quite a bit of agreement as well. With that in mind, removing the mention of `llms.txt`, as you suggested, seems like a good approach.

That said, if we simply remove that part, the clause starting with “which” would only refer to “offers tips for”, which feels a bit unbalanced. To address that, it might make sense to slightly expand the description and merge it with the following sentence, for example like this:

```suggestion
While 2025 has been widely described as the year of [vibe coding](https://x.com/karpathy/status/1886192184808149383), Astro has been steadily moving forward with support for AI tools as well. The Astro docs now include a new "[Build with AI](https://docs.astro.build/en/guides/build-with-ai/)" page that provides information about AI support on Discord and tips for developing Astro projects with AI, with a particular focus on the Astro Docs MCP Server, which is introduced below.
```

> While 2025 has been widely described as the year of [vibe coding](https://x.com/karpathy/status/1886192184808149383), Astro has been steadily moving forward with support for AI tools as well. The Astro docs now include a new "[Build with AI](https://docs.astro.build/en/guides/build-with-ai/)" page, which provides information about `llms.txt` and offers tips for developing Astro projects with AI. The most extensively covered topic on that page is the Astro Docs MCP Server, which is introduced below.

Author: Then I will go ahead and merge the second suggestion for now. I have also merged a few of the other suggestions you created. Really appreciate your feedback 🙏

- [Prerendered route collision warnings](https://astro.build/blog/astro-5140/#prerendered-route-collision-warnings)
- [SVG optimization with SVGO](https://astro.build/blog/astro-5160/#svg-optimization-with-svgo)

## Starlight

[Starlight](https://starlight.astro.build/) is an official theme for building documentation sites with Astro. Since it was released in 2023, it has been steadily gaining popularity, and in 2025 it was used for documentation by many different projects. There are too many to list them all here, but below are a few well known examples:

- [Nx](https://nx.dev/)
- [OpenCode](https://opencode.ai/)
- [Sharp](https://sharp.pixelplumbing.com/)
- [Gemini CLI](https://geminicli.com)

Starlight itself also continued to evolve throughout 2025, with updates including route data middleware, multi-site search support, automatic heading anchor links, and CSS cascade layers, among many others. To keep up with all the improvements, be sure to follow the [Starlight releases on GitHub](https://github.com/withastro/starlight/releases).

> 

Reviewer 2: > Sarah: I've also pulled Starlight out from the "Related Projects" section (since it's directly an Astro project, unlike the others) and I would love to add to the content and make that a more significant section. I've already asked some of the core Starlight team to have a look and help us figure out what more could go there, but if there's anything more you'd like to put there, Starlight has been a significant adoption path for many people using Astro (companies have been using Starlight for docs, THEN deciding to use Astro more and more). So I think a more substantial Starlight recap is certainly justified!

Here are some potential ideas from me that could add additional content (possibly sub-headings) to this page:

- Plugin API design changes in early 2025 -> lead to many new awesome plugins and existing plugins that got refactored to be better compatible with each other (also possible because of route data). Those topics also fit together, as they both (route data and plugin API change) got released in `0.32` (personally, I think this was one of the improvements with the biggest impact on possibilities / openness).
- Overall better compability and openness toward customization and usage: 
  - `0.33`: social link configuration for navbar: all built-in icons from Starlight can now be used
  - `0.35`: `generateId` allows custom slugs generated from file names
  - `0.35`: Use any built-in Starlight icons in Asides
  - `0.37`: Non-docs collection entries can use Starlight's processing pipeline.
- Automatic heading anchor links should have a dedicated paragraph I think.
- Cascade layers could also be expanded to it's own section.
- Many smaller accessibility improvements (fixing invalid meta tags, better system font stack, better legibility in tables thanks to changed `overflow-wrap` for Markdown content)
- A dedicated section to the awesome community around Starlight: New [content](https://starlight.astro.build/resources/community-content/), [plugins](https://starlight.astro.build/resources/plugins/), [themes](https://starlight.astro.build/resources/themes/) and [videos](https://starlight.astro.build/resources/community-content/#community-videos-and-streams) all over the place.

These are just my ideas, I am open for writing some sections if you need some help 👍

### GitHub Stars

Astro surpassed 50,000 GitHub stars in early 2025 and has now reached approximately 55,200. This represents an increase of nearly 7,000 stars since the beginning of 2025.

While 55,000 stars is already an impressive number, do you know where that places Astro among all repositories on GitHub? At the time of writing this article, only 295 repositories have more than 55,000 stars! Astro is ranked 293rd overall, placing it alongside projects such as [FFmpeg](https://github.com/FFmpeg/FFmpeg) and [OpenAI Codex](https://github.com/openai/codex). This shows that reaching 55,000 stars is a significant milestone.

Below is a [GitHub Star History](https://www.star-history.com/#withastro/astro) chart, which shows that Astro has steadily gained stars over the years.

<BlogContentImage src={starHistory} alt="A chart showing the increase of GitHub stars since the project's beginning (with zero stars) in 2021. The line tends up and to the right mostly linearly, with a slight jump in late 2022." />

### NPM Downloads

> Data from [npm trends](https://npmtrends.com/astro) shows that Astro had approximately 360,000 weekly downloads at the start of 2025, but this number grew to over 900,000 by the end of the year. This is about a two and a half times increase, showing that adoption has been accelerating compared to past years. This suggests that Astro is being adopted across a wide range of project types.

Reviewer 3: Feels weird to have "two and a half times" since we are using the numeral for all the other stats. Also, since it's English we should use a dot instead of a comma for the decimal separation.

```suggestion
Data from [npm trends](https://npmtrends.com/astro) shows that Astro had approximately 360.000 weekly downloads at the start of 2025, but this number grew to over 900.000 by the end of the year. This is about a 2.5x increase, showing that adoption has been accelerating compared to past years. This suggests that Astro is being adopted across a wide range of project types.
```

Astro was one of the fastest growing languages on GitHub in 2025!

<BlogContentImage src={octoverse} alt="Comparing August 2025 to August 2024, the fastest-growing languages by year-over-year growth are Luau (+194%), Typst (+108%), Astro (+78%), Blade (+67%), and TypeScript (+67%)." />

You may be thinking, "Wait, is Astro really a language?" That is a fair question, but if GitHub's language experts say so, who are we to argue? In any case, it's exciting to see Astro recognized in this way.

For more details, check out [the GitHub Octoverse 2025 website](https://github.blog/news-insights/octoverse/octoverse-a-new-developer-joins-github-every-second-as-ai-leads-typescript-to-1/#the-fastest-growing-languages-by-percentage-growth).

### Stack Overflow Developer Survey

[Astro is admired in the 2025 Stack Overflow Developer Survey](https://survey.stackoverflow.co/2025/technology#2-web-frameworks-and-technologies)!

This annual survey provides a snapshot into the needs of the global developer community, focusing on the tools and technologies they use or want to learn more about. Of all web frameworks and technologies, Astro was the 4th most admired at 62.2% of 23,678 respondents.

> <BlogContentImage src={stackOverflow} alt="Survey results of admired and desired web frameworks and technologies showing Astro as admired by 62.2% of respondents, behind only Phoenix (79%), Axum (76.4%) and Svelte (62.4%). Source: survey.stackoverflow.co/2025 Data licensed under Open Database License (ODbL)" />

Reviewer 3: ```suggestion
<BlogContentImage src={stackOverflow} alt="Survey results of admired and desired web frameworks and technologies showing Astro as admired by 62.2% of respondents, behind only Phoenix (79%), Axum (76.4%), and Svelte (62.4%). Source: survey.stackoverflow.co/2025 Data licensed under Open Database License (ODbL)" />
```

### JavaScript Rising Stars

[JavaScript Rising Stars](https://risingstars.js.org/2025/en) is an annual report that highlights the most popular and fast-growing projects in the JavaScript ecosystem based on the number of new GitHub stars gained over the past year.

> In the 2025 JavaScript Rising Stars, Astro ranked 4th in the Back-end/Full-stack category, and also placed 3rd in the Static Sites category. Additionally, Starlight ranked 8th in the Static Sites category. Even when compared with other projects, Astro stands out with strong recognition. Be sure to visit the site and check out the details.

Reviewer 3: ```suggestion
In the 2025 JavaScript Rising Stars, Astro ranked 4th in the Back-end/Full-stack category and also placed 3rd in the Static Sites category. Additionally, Starlight ranked 8th in the Static Sites category. Even when compared with other projects, Astro stands out with strong recognition. Be sure to visit the site and check out the details.
```

[Model Context Protocol](https://modelcontextprotocol.io/) (MCP) is a standard for AI tools to access external systems. With MCP servers, AI tools can perform actions such as searching documentation, running code, and other tasks.

The Astro Docs MCP server, which was released in July 2025, allows AI tools to access the official Astro documentation. By using this server as an information source, AI tools can avoid relying on outdated recommendations and understand current best practices. The process of connecting to an MCP server depends on the tool being used, and Astro documents getting started with several of the most popular editors, agents, and assistants.

[The official MCP inspector](https://modelcontextprotocol.io/docs/tools/inspector) is a convenient option to quickly try out the Astro Docs MCP server. You can launch it with the command `npx @modelcontextprotocol/inspector`, then choose **Streamable HTTP** as the Transport Type, and enter `https://mcp.docs.astro.build/mcp` as the URL.

> After you click **Connect**, the inspector will connect to the MCP server and the `search_astro_docs` tool will appear as an available item under **List Tools**. Selecting it and providing any Astro related topic as the query allows you to see the raw documentation data that the AI will receive:

Reviewer 3: ```suggestion
After you click **Connect**, the inspector will connect to the MCP server and the `search_astro_docs` tool will appear as an available item under **List Tools**. Selecting it and providing any Astro-related topic as the query allows you to see the raw documentation data that the AI will receive:
```

<BlogContentImage src={mcpInspector} alt="A screenshot of the MCP Inspector connected to the Astro Docs MCP Server, showing the 'search_astro_docs' tool and documentation data for 'Astro content collections'." />

[Check out the official docs for more information on the Astro Docs MCP Server](https://docs.astro.build/en/guides/build-with-ai/#astro-docs-mcp-server).

### `astro:config`

> In [v5.7.0](/blog/astro-570/#config-imports), Astro added support for importing a subset of the Astro configuration in a type safe way. The `astro:config` virtual module provides two submodules, `/client` and `/server`, which are designed to be used depending on the runtime environment.

Reviewer 3: ```suggestion
In [v5.7.0](/blog/astro-570/#config-imports), Astro added support for importing a subset of the Astro configuration in a type-safe way. The `astro:config` virtual module provides two submodules, `/client` and `/server`, which are designed to be used depending on the runtime environment.
```

- [Trailing slash redirects](/blog/astro-520/#trailing-slash-redirects)
- [Faster page rendering](/blog/astro-530/#faster-page-rendering)
- [Remote image optimization in Markdown](/blog/astro-540/#remote-image-optimization-in-markdown)
- [Experimental: preserve order of style and script tags](/blog/astro-550/#experimental-preserve-order-of-style-and-script-tags)
- [Prerendered route collision warnings](/blog/astro-5140/#prerendered-route-collision-warnings)
- [SVG optimization with SVGO](/blog/astro-5160/#svg-optimization-with-svgo)

## Starlight

[Starlight](https://starlight.astro.build/) is the official theme for building documentation sites with Astro. Since it was released in 2023, it has been steadily gaining popularity, and in 2025 Starlight powered documentation sites for [Cloudflare](https://developers.cloudflare.com/), [Google](https://geminicli.com/), [Microsoft](https://aspire.dev/), [Netlify](https://docs.netlify.com/), [OpenAI](https://openai.github.io/openai-agents-js/), [WPEngine](https://developers.wpengine.com/docs/), and countless others.

> Starlight itself also continued to evolve throughout 2025, seeing seven minor releases with updates including a powerful route data middleware system for implementing custom logic, multi-site search support, automatic heading anchor links, CSS cascade layers, and several accessibility improvements, among many others. Starlight ends 2025 with:

Reviewer 4: ```suggestion
Starlight itself also continued to evolve throughout the year, seeing seven minor releases with updates including a powerful route data middleware system for implementing custom logic, multi-site search support, automatic heading anchor links, CSS cascade layers, and several accessibility improvements, among many others. Starlight ends 2025 with:
```

## Looking Ahead to 2026 and Astro v6

Astro has a public [Project Roadmap](https://github.com/withastro/roadmap) where its future development direction is discussed. Here, we will take a brief look at some proposals at Stage 2 or above that are expected to be added to Astro, giving a preview of what is coming in 2026.

### Astro v6

While 2025 did not see a major version release, Astro v6 is right around the corner! An alpha version of v6 was first released in December 2025 with support for workerd in dev mode for Cloudflare, an update to Zod 4, the stabilization of many experimental features, and more. (See the [v6 branch CHANGELOG.md](https://github.com/withastro/astro/blob/next/packages/astro/CHANGELOG.md) for complete details.) Most of the v6 scope is already included in the alpha releases, and the first v6 beta release with an accompanying blog post is planned in the very near future!

 You can upgrade to the latest available v6 version using our [v6 Upgrade Guide](https://v6.docs.astro.build/en/guides/upgrade-to/v6/) which details all breaking changes that might require changes to your existing project code.

### Route Caching

> Route caching is an API that aims to provide declarative cache control based on web standards in a platform independent way. The early idea behind this new feature was first teased by <Mention name="matt" /> in [his ViteConf 2025 talk](#viteconf).

Reviewer 3: ```suggestion
Route caching is an API that aims to provide declarative cache control based on web standards in a platform-independent way. The early idea behind this new feature was first teased by <Mention name="matt" /> in [his ViteConf 2025 talk](#viteconf).
```

> Below is a sample taken from [the RFC](https://github.com/withastro/roadmap/pull/1245), which shows that you can configure values such as `maxAge`, `swr`, and `tags` on a per route basis, and invalidate caches by `path` or `tag`. It is also designed to support cache invalidation in combination with the `cacheHints` returned from `getLiveEntry()` in the Live Content Collections mentioned earlier:

Reviewer 3: ```suggestion
Below is a sample taken from [the RFC](https://github.com/withastro/roadmap/pull/1245), which shows that you can configure values such as `maxAge`, `swr`, and `tags` on a per-route basis, and invalidate caches by `path` or `tag`. It is also designed to support cache invalidation in combination with the `cacheHints` returned from `getLiveEntry()` in the Live Content Collections mentioned earlier:
```

  // Invalidate by path
  cache.invalidate({ path: "/products/laptop" });

  return Response.json({ ok: true });
};
```

```ts title="Cache invalidation by tag"
// src/pages/api/webhook.ts
export const POST: APIRoute = ({ cache }) => {
  // Invalidate by tag
  cache.invalidate({ tag: "products" });

  return Response.json({ ok: true });
};
```

```ts title="Integration with live collections"
---
// src/pages/products/[slug].astro
import { getLiveEntry, render } from 'astro:content';

const { entry, cacheHint } = await getLiveEntry('products', Astro.params.slug);

// Apply cache hints from the live collection loader
Astro.cache(cacheHint);

const { Content } = await render(entry);
---
<h1>{entry.data.name}</h1>
<Content />
```

Although this API is not finalized yet, it is great to see how a declarative approach can make configuration shorter and easier to understand compared to manual setup. It is also helpful that different cache invalidation mechanisms across platforms can be unified behind a single API. Once this lands in Astro, more complex caching strategies such as ISR should become much easier to implement.

### Tracing Hooks for Astro

> This is a proposal to add a low overhead, general purpose instrumentation hook system to the Astro core.

Reviewer 3: ```suggestion
This is a proposal to add a low-overhead, general-purpose instrumentation hook system to the Astro core.
```

## Closing Thoughts

Whew, that was a long one. Thank you so much for reading through to the end!

Looking back, I hope this article conveyed just how much the Astro ecosystem has matured. Further enhancements to existing features like Content Collections, as well as the introduction of thoughtful APIs for traditionally difficult areas such as responsive images and fonts, feel very much in line with Astro philosophy. Content caching is another area that often causes pain for developers due to its close ties to infrastructure, so I am particularly excited about the APIs aiming to address those challenges.

I cannot wait to see what new features and improvements Astro brings in 2026.

<Note title="About the author">
Hey, my name is <Mention name="shinya" />. I'm a web developer living in Tokyo, Japan. I'm interested in what the future of the web will be like and I would like to be a part of it in some way.

> </Note>

Reviewer 5: This diff looks weird @yanthomasdev , it seems you're replacing the whole post with a single paragraph 😱 (at least on my end. Probably a GitHub new experience bug and this was only `l123`? 😅 )
